src/common/adapters/MultiAdapter.sol

```diff
@@ -0,0 +1,239 @@
+            }
+        } else if (!isMessageProof) {
+            state.pending = payload;
+        }
+    }
+
+    /// @inheritdoc IMultiAdapter
+    function initiateRecovery(uint16 centrifugeId, IAdapter adapter, bytes32 payloadHash) external auth {
+        require(_adapterDetails[centrifugeId][adapter].id != 0, InvalidAdapter());
+        recoveries[centrifugeId][adapter][payloadHash] = block.timestamp + RECOVERY_CHALLENGE_PERIOD;
+        emit InitiateRecovery(centrifugeId, payloadHash, adapter);
+    }
+
+    /// @inheritdoc IMultiAdapter
+    function disputeRecovery(uint16 centrifugeId, IAdapter adapter, bytes32 payloadHash) external auth {
+        delete recoveries[centrifugeId][adapter][payloadHash];
+        emit DisputeRecovery(centrifugeId, payloadHash, adapter);
+    }
+
+    /// @inheritdoc IMultiAdapter
+    function executeRecovery(uint16 centrifugeId, IAdapter adapter, bytes calldata payload) external {
+        bytes32 payloadHash = keccak256(payload);
+        uint256 recovery = recoveries[centrifugeId][adapter][payloadHash];
+
+        require(recovery != 0, RecoveryNotInitiated());
+        require(recovery <= block.timestamp, RecoveryChallengePeriodNotEnded());
+
+        delete recoveries[centrifugeId][adapter][payloadHash];
+        _handle(centrifugeId, payload, adapter);
+        emit ExecuteRecovery(centrifugeId, payload, adapter);
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Outgoing
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IAdapter
+    function send(uint16 centrifugeId, bytes calldata payload, uint256 gasLimit, address refund)
+        external
+        payable
+        auth
+        returns (bytes32)
+    {
+        IAdapter[] memory adapters_ = adapters[centrifugeId];
+        require(adapters_.length != 0, EmptyAdapterSet());
+
+        bytes32 payloadHash = keccak256(payload);
+        bytes32 payloadId = keccak256(abi.encodePacked(localCentrifugeId, centrifugeId, payloadHash));
+        bytes memory proof = payloadHash.serializeMessageProof();
+
+        uint256 cost = adapters_[0].estimate(centrifugeId, payload, gasLimit);
+        bytes32 adapterData = adapters_[0].send{value: cost}(centrifugeId, payload, gasLimit, refund);
+        emit SendPayload(centrifugeId, payloadId, payload, adapters_[0], adapterData, refund);
+
+        for (uint256 i = 1; i < adapters_.length; i++) {
+            cost = adapters_[i].estimate(centrifugeId, proof, gasLimit);
+            adapterData = adapters_[i].send{value: cost}(centrifugeId, proof, gasLimit, refund);
+            emit SendProof(centrifugeId, payloadId, payloadHash, adapters_[i], adapterData);
+        }
+
+        return bytes32(0);
+    }
+
+    /// @inheritdoc IAdapter
+    function estimate(uint16 centrifugeId, bytes calldata payload, uint256 gasLimit)
+        public
```

Comment:
This can be `external`

Reply:
Correct. I apply this to Axelar and Wormhole adapters